Doesn't build with yaml-cpp 0.6.0

Attempted build on Fedora Rawhide with GCC 8 preview:

`
[ 73%] Linking CXX executable ocio_core_tests
cd /builddir/build/BUILD/OpenColorIO-1.1.0/build/src/core_tests && /usr/bin/cmake -E cmake_link_script CMakeFiles/ocio_core_tests.dir/link.txt --verbose=1
/usr/bin/c++  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -mcet -fcf-protection -Wall -Wextra -Wshadow -Wconversion -Wcast-qual -Wformat=2 -msse2 -O2 -g -DNDEBUG  -Wl,-z,relro  -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -rdynamic CMakeFiles/ocio_core_tests.dir/__/core/AllocationOp.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/AllocationTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Baker.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/CDLTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Caching.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/ColorSpace.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/ColorSpaceTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Config.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Context.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Display.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/DisplayTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Exception.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/ExponentOps.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/ExponentTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormat3DL.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatCC.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatCCC.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatCDL.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatCSP.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatHDL.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatIridasCube.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatIridasItx.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatIridasLook.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatPandora.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatSpi1D.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatSpi3D.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatSpiMtx.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatTruelight.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileFormatVF.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/FileTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/GpuShaderDesc.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/GpuShaderUtils.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/GroupTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/HashUtils.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/ImageDesc.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/ImagePacking.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/LogOps.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/LogTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Logging.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Look.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/LookParse.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/LookTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Lut1DOp.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Lut3DOp.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/MathUtils.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/MatrixOps.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/MatrixTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/NoOps.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/OCIOYaml.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Op.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/OpOptimizers.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/ParseUtils.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/PathUtils.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Platform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Processor.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/ScanlineHelper.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/Transform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/TruelightOp.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/TruelightTransform.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/UnitTest.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/md5/md5.cpp.o CMakeFiles/ocio_core_tests.dir/__/core/pystring/pystring.cpp.o  -o ocio_core_tests -ltinyxml -lyaml-cpp 
make[2]: Leaving directory '/builddir/build/BUILD/OpenColorIO-1.1.0/build'
CMakeFiles/ocio_core_tests.dir/__/core/OCIOYaml.cpp.o: In function `YAML::Exception::Exception(YAML::Mark const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/include/yaml-cpp/exceptions.h:122: undefined reference to `vtable for YAML::Exception'
CMakeFiles/ocio_core_tests.dir/__/core/OCIOYaml.cpp.o: In function `YAML::RepresentationException::RepresentationException(YAML::Mark const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/include/yaml-cpp/exceptions.h:155: undefined reference to `vtable for YAML::RepresentationException'
CMakeFiles/ocio_core_tests.dir/__/core/OCIOYaml.cpp.o: In function `YAML::Exception::Exception(YAML::Mark const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/include/yaml-cpp/exceptions.h:122: undefined reference to `vtable for YAML::Exception'
CMakeFiles/ocio_core_tests.dir/__/core/OCIOYaml.cpp.o: In function `YAML::RepresentationException::RepresentationException(YAML::Mark const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/include/yaml-cpp/exceptions.h:155: undefined reference to `vtable for YAML::RepresentationException'
CMakeFiles/ocio_core_tests.dir/__/core/OCIOYaml.cpp.o: In function `YAML::Exception::Exception(YAML::Mark const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/include/yaml-cpp/exceptions.h:122: undefined reference to `vtable for YAML::Exception'
CMakeFiles/ocio_core_tests.dir/__/core/OCIOYaml.cpp.o: In function `YAML::RepresentationException::RepresentationException(YAML::Mark const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/include/yaml-cpp/exceptions.h:155: undefined reference to `vtable for YAML::RepresentationException'
/usr/bin/ld: ocio_core_tests: hidden symbol `_ZTVN4YAML9ExceptionE' isn't defined
/usr/bin/ld: final link failed: Bad value
`

I disabled tests and get a similar error while linking the main library, but it's not QUITE the same:

`
[ 63%] Linking CXX shared library libOpenColorIO.so
cd /builddir/build/BUILD/OpenColorIO-1.1.0/build/src/core && /usr/bin/cmake -E cmake_link_script CMakeFiles/OpenColorIO.dir/link.txt --verbose=1
/usr/bin/c++ -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -mcet -fcf-protection -Wall -Wextra -Wshadow -Wconversion -Wcast-qual -Wformat=2 -msse2 -O2 -g -DNDEBUG -Wl,-z,relro  -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -shared -Wl,-soname,libOpenColorIO.so.1 -o libOpenColorIO.so.1.1.0 CMakeFiles/OpenColorIO.dir/AllocationOp.cpp.o CMakeFiles/OpenColorIO.dir/AllocationTransform.cpp.o CMakeFiles/OpenColorIO.dir/Baker.cpp.o CMakeFiles/OpenColorIO.dir/CDLTransform.cpp.o CMakeFiles/OpenColorIO.dir/Caching.cpp.o CMakeFiles/OpenColorIO.dir/ColorSpace.cpp.o CMakeFiles/OpenColorIO.dir/ColorSpaceTransform.cpp.o CMakeFiles/OpenColorIO.dir/Config.cpp.o CMakeFiles/OpenColorIO.dir/Context.cpp.o CMakeFiles/OpenColorIO.dir/Display.cpp.o CMakeFiles/OpenColorIO.dir/DisplayTransform.cpp.o CMakeFiles/OpenColorIO.dir/Exception.cpp.o CMakeFiles/OpenColorIO.dir/ExponentOps.cpp.o CMakeFiles/OpenColorIO.dir/ExponentTransform.cpp.o CMakeFiles/OpenColorIO.dir/FileFormat3DL.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatCC.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatCCC.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatCDL.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatCSP.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatHDL.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatIridasCube.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatIridasItx.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatIridasLook.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatPandora.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatSpi1D.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatSpi3D.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatSpiMtx.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatTruelight.cpp.o CMakeFiles/OpenColorIO.dir/FileFormatVF.cpp.o CMakeFiles/OpenColorIO.dir/FileTransform.cpp.o CMakeFiles/OpenColorIO.dir/GpuShaderDesc.cpp.o CMakeFiles/OpenColorIO.dir/GpuShaderUtils.cpp.o CMakeFiles/OpenColorIO.dir/GroupTransform.cpp.o CMakeFiles/OpenColorIO.dir/HashUtils.cpp.o CMakeFiles/OpenColorIO.dir/ImageDesc.cpp.o CMakeFiles/OpenColorIO.dir/ImagePacking.cpp.o CMakeFiles/OpenColorIO.dir/LogOps.cpp.o CMakeFiles/OpenColorIO.dir/LogTransform.cpp.o CMakeFiles/OpenColorIO.dir/Logging.cpp.o CMakeFiles/OpenColorIO.dir/Look.cpp.o CMakeFiles/OpenColorIO.dir/LookParse.cpp.o CMakeFiles/OpenColorIO.dir/LookTransform.cpp.o CMakeFiles/OpenColorIO.dir/Lut1DOp.cpp.o CMakeFiles/OpenColorIO.dir/Lut3DOp.cpp.o CMakeFiles/OpenColorIO.dir/MathUtils.cpp.o CMakeFiles/OpenColorIO.dir/MatrixOps.cpp.o CMakeFiles/OpenColorIO.dir/MatrixTransform.cpp.o CMakeFiles/OpenColorIO.dir/NoOps.cpp.o CMakeFiles/OpenColorIO.dir/OCIOYaml.cpp.o CMakeFiles/OpenColorIO.dir/Op.cpp.o CMakeFiles/OpenColorIO.dir/OpOptimizers.cpp.o CMakeFiles/OpenColorIO.dir/ParseUtils.cpp.o CMakeFiles/OpenColorIO.dir/PathUtils.cpp.o CMakeFiles/OpenColorIO.dir/Platform.cpp.o CMakeFiles/OpenColorIO.dir/Processor.cpp.o CMakeFiles/OpenColorIO.dir/ScanlineHelper.cpp.o CMakeFiles/OpenColorIO.dir/Transform.cpp.o CMakeFiles/OpenColorIO.dir/TruelightOp.cpp.o CMakeFiles/OpenColorIO.dir/TruelightTransform.cpp.o CMakeFiles/OpenColorIO.dir/UnitTest.cpp.o CMakeFiles/OpenColorIO.dir/md5/md5.cpp.o CMakeFiles/OpenColorIO.dir/pystring/pystring.cpp.o -ltinyxml -lyaml-cpp 
make[2]: Leaving directory '/builddir/build/BUILD/OpenColorIO-1.1.0/build'
CMakeFiles/OpenColorIO.dir/OCIOYaml.cpp.o: In function `YAML::Exception::Exception(YAML::Mark const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/include/yaml-cpp/exceptions.h:122: undefined reference to `vtable for YAML::Exception'
/usr/bin/ld: CMakeFiles/OpenColorIO.dir/OCIOYaml.cpp.o: relocation R_X86_64_PC32 against undefined hidden symbol `_ZTVN4YAML9ExceptionE' can not be used when making a shared object
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
`

An undefined reference to the `vtable` typically means that you are declaring a virtual method without actually defining it. g++ can then (based on factors that only the compiler authors really understand) decide to pick that method as the key method for emitting the `vtable` and you are out of luck. So fix that: any declared virtual method **must** be implemented.

--- a/share/cmake/modules/FindYamlCpp.cmake
+++ b/share/cmake/modules/FindYamlCpp.cmake
@@ -93,15 +93,17 @@ if(NOT OCIO_INSTALL_EXT_PACKAGES STREQUAL ALL)
     )
 
     # Get version from config if it was found.
-    # Important: The yaml-cpp interface changed between 0.3.0 and 0.5.0. 
-    # OCIO supports both, but relies on YAMLCPP_VERSION to set the OLDYAML
-    # property below, which tells the compiler which to use.
     if(PC_YAMLCPP_VERSION)
         set(YAMLCPP_VERSION "${PC_YAMLCPP_VERSION}")
     elseif(EXISTS "${YAMLCPP_INCLUDE_DIR}/yaml-cpp/iterator.h")
         set(YAMLCPP_VERSION "0.3.0")
+    elseif(EXISTS "${YAMLCPP_INCLUDE_DIR}/yaml-cpp/noncopyable.h")
+        # The version is higher than 0.3.0 but lower than 0.6.3
+        # i.e. the version 0.6.3 removes the file 'yaml-cpp/noncopyable.h.
+        set(YAMLCPP_VERSION "0.5.3")
     else()
-        set(YAMLCPP_VERSION "0.5.0")
+        # The only supported version.
+        set(YAMLCPP_VERSION "0.6.3")
     endif()
 
     # Override REQUIRED if package can be installed

--- a/src/OpenColorIO/Config.cpp
+++ b/src/OpenColorIO/Config.cpp
@@ -227,8 +227,6 @@ OCIO_NAMESPACE_ENTER
         mutable StringMap cacheids_;
         mutable std::string cacheidnocontext_;
         
-        OCIOYaml io_;
-        
         Impl() : 
             majorVersion_(FirstSupportedMajorVersion_),
             minorVersion_(0),
@@ -335,10 +333,8 @@ OCIO_NAMESPACE_ENTER
     {
         std::istringstream istream;
         istream.str(INTERNAL_RAW_PROFILE);
-        
-        ConfigRcPtr config = Config::Create();
-        config->getImpl()->io_.open(istream, config);
-        return config;
+
+        return CreateFromStream(istream);
     }
 
     ConstConfigRcPtr Config::CreateFromEnv()
@@ -364,16 +360,16 @@ OCIO_NAMESPACE_ENTER
             os << "' OCIO profile.";
             throw Exception (os.str().c_str());
         }
-        
+
         ConfigRcPtr config = Config::Create();
-        config->getImpl()->io_.open(istream, config, filename);
+        OCIOYaml::read(istream, config, filename);
         return config;
     }
     
     ConstConfigRcPtr Config::CreateFromStream(std::istream & istream)
     {
         ConfigRcPtr config = Config::Create();
-        config->getImpl()->io_.open(istream, config);
+        OCIOYaml::read(istream, config, nullptr);
         return config;
     }
     
@@ -1624,7 +1620,7 @@ OCIO_NAMESPACE_ENTER
     {
         try
         {
-            getImpl()->io_.write(os, this);
+            OCIOYaml::write(os, this);
         }
         catch( const std::exception & e)
         {

--- a/src/OpenColorIO/OCIOYaml.cpp
+++ b/src/OpenColorIO/OCIOYaml.cpp
@@ -5,57 +5,7 @@
 
 #include <OpenColorIO/OpenColorIO.h>
 
-#ifndef _WIN32
-
-// fwd declare yaml-cpp visibility
-#pragma GCC visibility push(hidden)
-namespace YAML {
-    class Exception;
-    class BadDereference;
-    class RepresentationException;
-    class EmitterException;
-    class ParserException;
-    class InvalidScalar;
-    class KeyNotFound;
-    template <typename T> class TypedKeyNotFound;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::ColorSpace>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::Config>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::Exception>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::GpuShaderDesc>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::ImageDesc>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::Look>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::Processor>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::Transform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::AllocationTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::CDLTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::ColorSpaceTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::DisplayTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::ExponentTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::ExponentWithLinearTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::ExposureContrastTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::FileTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::FixedFunctionTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::GroupTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::LogAffineTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::LogTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::LookTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::MatrixTransform>;
-    template <> class TypedKeyNotFound<OCIO_NAMESPACE::RangeTransform>;
-}
-#pragma GCC visibility pop
-
-#endif
-
-#ifdef _MSC_VER
-#pragma warning( push )
-#pragma warning( disable: 4146 )
-#endif
-
 #include <yaml-cpp/yaml.h>
-
-#ifdef _MSC_VER
-#pragma warning( pop )
-#endif
 
 #include "Display.h"
 #include "Logging.h"
@@ -2481,7 +2431,7 @@ OCIO_NAMESPACE_ENTER
     
     ///////////////////////////////////////////////////////////////////////////
     
-    void OCIOYaml::open(std::istream& istream, ConfigRcPtr& c, const char* filename) const
+    void OCIOYaml::read(std::istream & istream, ConfigRcPtr & c, const char * filename)
     {
         try
         {
@@ -2498,7 +2448,7 @@ OCIO_NAMESPACE_ENTER
         }
     }
     
-    void OCIOYaml::write(std::ostream& ostream, const Config* c) const
+    void OCIOYaml::write(std::ostream & ostream, const Config * c)
     {
         YAML::Emitter out;
         out.SetDoublePrecision(std::numeric_limits<double>::digits10);

--- a/src/OpenColorIO/OCIOYaml.h
+++ b/src/OpenColorIO/OCIOYaml.h
@@ -9,11 +9,10 @@
 OCIO_NAMESPACE_ENTER
 {
     
-    class OCIOYaml
+    namespace OCIOYaml
     {
-    public:
-        void open(std::istream& istream, ConfigRcPtr& c, const char* filename = NULL) const;
-        void write(std::ostream& ostream, const Config* c) const;
+        void read(std::istream & istream, ConfigRcPtr & c, const char * filename);
+        void write(std::ostream & ostream, const Config * c);
     };
     
 }

